jurigged/codedb.py

```diff
@@ -0,0 +1,156 @@
+import gc
+import sys
+import time
+import types
+
+from ovld import ovld
+
+##########
+# CodeDB #
+##########
+
+MAX_TIME = 0.1
+
+
+def make_audithook(self):
+    def watch_exec(event, obj):  # pragma: no cover
+        # Note: Python does not trace audit hooks, so normal use will not show
+        # coverage of this function even if it is executed
+        if event == "exec":
+            (code,) = obj
+            self.assimilate(code)
+
+    # The closure is faster than a method on _CodeDB
+    return watch_exec
+
+
+class _CodeDB:
+    def __init__(self):
+        self.codes = {}
+        self.functions = {}
+        self.last_cost = 0
+        self.always_use_cache = False
+
+    def setup(self):
+        self.collect_all()
+        sys.addaudithook(make_audithook(self))
+
+    def collect_all(self):
+        results = []
+        for obj in gc.get_objects():
+            if isinstance(obj, types.FunctionType):
+                results.append((obj, obj.__code__))
+            elif not hasattr(obj, "__getattr__"):
```

Comment:
If you're trying to avoid side effects from checking for attributes, I suggest `inspect.getattr_static`

Reply:
Ahh, I didn't know that method existed. That'll be handy, thanks.